src/replay_stage.rs

```diff
@@ -194,76 +186,155 @@ impl ReplayStage {
         }
 
         *entry_height.write().unwrap() += entries_len;
+
         res?;
         inc_new_counter_info!(
             "replicate_stage-duration",
             duration_as_ms(&now.elapsed()) as usize
         );
-
         Ok(())
     }
 
     #[allow(clippy::new_ret_no_self, clippy::too_many_arguments)]
     pub fn new(
         my_id: Pubkey,
         voting_keypair: Option<Arc<VotingKeypair>>,
+        db_ledger: Arc<DbLedger>,
         bank: Arc<Bank>,
         cluster_info: Arc<RwLock<ClusterInfo>>,
-        window_receiver: EntryReceiver,
         exit: Arc<AtomicBool>,
         entry_height: Arc<RwLock<u64>>,
         last_entry_id: Arc<RwLock<Hash>>,
         to_leader_sender: TvuRotationSender,
         entry_stream: Option<&String>,
+        ledger_signal_sender: SyncSender<bool>,
+        ledger_signal_receiver: Receiver<bool>,
     ) -> (Self, EntryReceiver) {
         let (ledger_entry_sender, ledger_entry_receiver) = channel();
         let mut entry_stream = entry_stream.cloned().map(EntryStream::new);
 
+        let (_, mut current_slot) = bank
+            .get_current_leader()
+            .expect("Scheduled leader should be calculated by this point");
+
+        let mut max_tick_height_for_slot = bank
+            .leader_scheduler
+            .read()
+            .unwrap()
+            .max_tick_height_for_slot(current_slot);
+
+        let exit_ = exit.clone();
         let t_replay = Builder::new()
             .name("solana-replay-stage".to_string())
             .spawn(move || {
-                let _exit = Finalizer::new(exit);
-                let entry_height_ = entry_height;
-                let last_entry_id = last_entry_id;
+                let _exit = Finalizer::new(exit_.clone());
                 let (mut last_leader_id, _) = bank
                     .get_current_leader()
                     .expect("Scheduled leader should be calculated by this point");
-                loop {
-                    let (leader_id, _) = bank
-                        .get_current_leader()
-                        .expect("Scheduled leader should be calculated by this point");
-                    if leader_id != last_leader_id && leader_id == my_id {
-                        to_leader_sender
-                            .send(TvuReturnType::LeaderRotation(
-                                bank.tick_height(),
-                                *entry_height_.read().unwrap(),
-                                *last_entry_id.read().unwrap(),
-                            ))
-                            .unwrap();
+                'outer: loop {
```

Comment:
Ouch, having to break out of an outer loop like this is usually a sign of unwanted complexity in the implementation.   Can you refactor to avoid this?